feat(adapters): Drizzle adapter

anthonyshew opened the pull request:
## ☕️ Reasoning

Introducing a `drizzle-orm` adapter!

With `drizzle`'s increasing popularity, it's no surprise that [there's a discussion](https://github.com/nextauthjs/next-auth/discussions/7005) for an adapter. This PR adds it.

## 🧢 Checklist

- [x] Documentation
- [x] Tests
- [x] Ready to be merged

## 📌 Resources

- [Security guidelines](https://github.com/nextauthjs/.github/blob/main/SECURITY.md)
- [Contributing guidelines](https://github.com/nextauthjs/.github/blob/main/CONTRIBUTING.md)
- [Code of conduct](https://github.com/nextauthjs/.github/blob/main/CODE_OF_CONDUCT.md)
- [Contributing to Open Source](https://kcd.im/pull-request)

# Notes
- There is currently an open issue with Drizzle's recent versions that breaks monorepo dependencies. Please use 0.24.x until this issue is resolved: https://github.com/drizzle-team/drizzle-orm/issues/283
- There is currently an open issue with Drizzle truncating millisecond precision. I've added an extra option to the `runBasicTests` utility to be able to provide some custom fixtures to workaround this. https://github.com/drizzle-team/drizzle-orm/pull/668

balazsorban44 committed:
chore: bump turbo and pnpm

balazsorban44 committed:
chore: bump turbo and pnpm

jakzo committed:
chore(docs): fix dynamodb typo (#7130)

fix: typo

balazsorban44 committed:
chore: bump pnpm

balazsorban44 committed:
chore: bump pnpm

balazsorban44 committed:
Merge branch 'main' of github.com:nextauthjs/next-auth

balazsorban44 committed:
chore: update lock file, bump dev dependencies

balazsorban44 committed:
chore: update lock file, bump dev dependencies

balazsorban44 committed:
chore: run `pnpm install --fix-lockfile`

balazsorban44 committed:
chore: re-run pnpm install

balazsorban44 committed:
chore: add missing dev dep

balazsorban44 committed:
chore: add missing dev dep

SauravMaheshkar committed:
chore: move `prettier` and `eslint` configs under `package.json` (#7145)

balazsorban44 committed:
chore: fix lock file

balazsorban44 committed:
Merge branch 'main' of github.com:nextauthjs/next-auth

balazsorban44 committed:
chore: bump prisma dev version

balazsorban44 committed:
chore: format `package.json`

balazsorban44 committed:
chore: fix msw

balazsorban44 committed:
chore: revert

balazsorban44 committed:
chore: revert

anthonyshew committed:
Drizzle adapter!

anthonyshew committed:
Drizzle adapter!

balazsorban44 committed:
Update pnpm-lock.yaml

balazsorban44 committed:
Update pnpm-lock.yaml

balazsorban44 committed:
Update pnpm-lock.yaml

balazsorban44 committed:
Update pnpm-lock.yaml

balazsorban44 committed:
Update package.json

balazsorban44 committed:
Update package.json

balazsorban44 committed:
Update pnpm-lock.yaml

balazsorban44 committed:
Update pnpm-lock.yaml

balazsorban44 committed:
Update package.json

balazsorban44 committed:
Update package.json

balazsorban44 committed:
Update package.json

balazsorban44 committed:
Update package.json

balazsorban44 committed:
Update package.json

balazsorban44 committed:
Update package.json

balazsorban44 committed:
Update package.json

balazsorban44 committed:
Update package.json

balazsorban44 committed:
Update package.json

balazsorban44 committed:
Update package.json

balazsorban44 committed:
Update package.json

balazsorban44 committed:
Update package.json

balazsorban44 committed:
Update package.json

balazsorban44 committed:
Update package.json

balazsorban44 committed:
Update pnpm-lock.yaml

balazsorban44 committed:
Update pnpm-lock.yaml

balazsorban44 committed:
Update pnpm-lock.yaml

balazsorban44 committed:
Update pnpm-lock.yaml

balazsorban44 committed:
Update package.json

balazsorban44 committed:
Update package.json

balazsorban44 committed:
Update package.json

balazsorban44 committed:
Update package.json

balazsorban44 committed:
Update pnpm-lock.yaml

balazsorban44 committed:
Update pnpm-lock.yaml

balazsorban44 committed:
Update index.ts

balazsorban44 committed:
Update index.ts

balazsorban44 committed:
Update pnpm-lock.yaml

balazsorban44 committed:
Update pnpm-lock.yaml

balazsorban44 committed:
Delete pnpm-lock.yaml

balazsorban44 committed:
Delete pnpm-lock.yaml

balazsorban44 commented:
Good point in https://github.com/nextauthjs/next-auth/discussions/7005#discussioncomment-5542019, can we make sure to support multiple SQL flavors? :thinking:

anthonyshew committed:
Now fix TypeScript.

anthonyshew committed:
Now fix TypeScript.

anthonyshew committed:
Implementations, I think.

anthonyshew committed:
Implementations, I think.

anthonyshew committed:
Right command for test files.

anthonyshew committed:
Right command for test files.

anthonyshew commented:
@infamous55 Correct! That was a first cut at making things go. I meant to mark it as a draft but forgot to hit the button. 😅

I've got the different flavors on there now!

infamous55 commented:
@anthonyshew Glad to see you added support for the other SQL flavors!

I would also like to suggest having only one Drizzle adapter function that returns the correct type of adapter depending on the database client it receives. Like this:

```ts
export function DrizzleAdapter(db: any): Adapter {
	if (isMySqlDatabase(db)) return MySqlAdapter(db)
	else if (isPgDatabase(db)) return PgAdapter(db)
	else if (isSQLiteDatabase(db)) return SQLiteAdapter(db)
	else throw new Error("Unsupported database")
}
```

```ts
import { MySqlDatabase } from "drizzle-orm/mysql-core"
import { PgDatabase } from "drizzle-orm/pg-core"
import { BaseSQLiteDatabase } from "drizzle-orm/sqlite-core"

export function isMySqlDatabase(db: any): db is AnyMySqlDatabase {
	return db instanceof MySqlDatabase<any, any>
}

export function isPgDatabase(db: any): db is AnyPgDatabase {
	return db instanceof PgDatabase<any, any>
}
	 
export function isSQLiteDatabase(db: any): db is AnySQLiteDatabase {
	return db instanceof BaseSQLiteDatabase<any, any>
}
```

In this case, the user would only have to import `DrizzleAdapter` from `@next-auth/adapter-drizzle`. I find this setup the cleanest, but this is my personal opinion. It's up to you to decide how you want the publishing done.

anthonyshew committed:
Add compatibility layer adapter.

anthonyshew committed:
Add compatibility layer adapter.

anthonyshew commented:
@infamous55 I had considered this as well!

I didn't know how much we cared about compiled size so I did them split up. API convenience vs. bundle sizing tradeoff, a decision I'll let the core maintainers make. 😄 

I've now added another adapter directory (`adapter-drizzle-all`) using your strategy for comparison.

I'll attach bundle sizes below:

| SQLite | Postgres | MySQL | All 3 |
|--------|--------|--------|--------|
| 28k | 28k | 36k | 120k |

Still incredibly small, of course, but don't know how strict about bundle sizes @balazsorban44 wants to be!

To summarize, where this PR currently stands:
- Four adapters currently written (one for each flavor, one with a compatibility later for all three)
- Tests in `adapter-drizzle-sqlite` work
- Documentation is not handled (waiting on API surface decision and then happy to write!)

jeanhdev commented:
Hey guys - I'd really like to use the adapter for Drizzle with my PSQL DB, any news on this ?

rikimbili commented:
> Hey guys - I'd really like to use the adapter for Drizzle with my PSQL DB, any news on this ?

As a temporary or permanent solution you can create your own adapter [using the code here from this PR](https://github.com/anthonyshew/next-auth/blob/main/packages/adapter-drizzle-pg/src/index.ts). This way you can also modify how user data is stored and queried from your db if you have something more custom. I tried this approach a week ago and the adapter has been working great.

vilius commented on packages/adapter-drizzle-all/src/sqlite/schema.ts:
```diff
@@ -0,0 +1,52 @@
+  id: text('id').notNull().primaryKey(),
+  name: text('name'),
+  email: text("email").notNull(),
+  emailVerified: integer("emailVerified", { mode: "timestamp_ms" }),
```
`emailVerifiedAt` would be more consistent

ferntheplant commented:
> As a temporary or permanent solution you can create your own adapter [using the code here from this PR](https://github.com/anthonyshew/next-auth/blob/main/packages/adapter-drizzle-pg/src/index.ts)

Whenever my `user` model extends the minimal `id, name, email, image` fields I get some pretty silly type errors. For example `adapter.createUser` **exclusively** accepts an object of shape 
```typescript
type Omit<AdapterUser, "id"> = { name?: string | null, email?: string | null, image?: string | null }```

I think it really should accept anything that _extends_ this type instead so I can extend my user schema in a typesafe way

ferntheplant commented on packages/adapter-drizzle-all/src/pg/index.ts:
```diff
@@ -0,0 +1,264 @@
+ * export const sessions = pgTable("sessions", {
+ *  userId: text("userId").notNull().references(() => users.id, { onDelete: "cascade" }),
+ *  sessionToken: text("sessionToken").notNull().primaryKey(),
+ *  expires: integer("expires").notNull(),
```
this and the verificationTokens expires below should be timestamps not integers

anthonyshew replied:
I'll make sure docs get handled after interfaces are settled! https://github.com/nextauthjs/next-auth/pull/7165#issuecomment-1500571782

vixeven committed:
fix(docs): fix default `maxAge` formula (#7406)

balazsorban44 committed:
feat(adapters): add Account mapping before database write (#7369)

* feat: map Account before saving to database

* document `acconut()`, explain default behaviour

* generate `expires_at` based on `expires_in`

Fixes #6538

* rename

* strip undefined on `defaultProfile`

* don't forward defaults to account callback

* improve internal namings, types, docs

balazsorban44 committed:
chore: improve errors, add more docs (#7415)

* JWT Token -> JWT

* document some errors

* improve errors, docs

balazsorban44 committed:
fix: loosen profile types

balazsorban44 committed:
chore: type fixes

balazsorban44 committed:
fix: allow handling OAuth callback error response

related #7407

pyrossh commented on packages/adapter-drizzle-all/src/mysql/index.ts:
```diff
@@ -0,0 +1,293 @@
+    createUser: async (data) => {
+      await client
+        .insert(users)
+        .values({ ...data, id: "123" })
```
Can these mocked id's be removed?

pyrossh commented on packages/adapter-drizzle-all/src/pg/schema.ts:
```diff
@@ -0,0 +1,48 @@
+import { integer, timestamp, pgTable, text, primaryKey } from 'drizzle-orm/pg-core';
+import { drizzle } from 'drizzle-orm/postgres-js';
```
There are serverless drivers as well that needs to be taken into consideration. Ex: I use `drizzle-orm/neon-serverless` driver for postgres instances hosted by neon. More drivers here https://github.com/drizzle-team/drizzle-orm/blob/main/drizzle-orm/src/pg-core/README.md

anthonyshew replied:
I'm under the impression that the drivers are handled in userland code.

Users will pass their final database client to `next-auth` which will use whichever SQL flavor the user's driver is using. As long as your SQL flavor matches the `next-auth` adapter you're using, you should be good to go.

That's at least my impression of things; hope I'm not being a fool. 😀

vilius commented on packages/adapter-drizzle-all/src/mysql/index.ts:
```diff
@@ -0,0 +1,293 @@
+    },
+    getUserByAccount: async (account) => {
+      return client.select({
+        id: users.id,
```
Inconsistency: `getUser` and `getUserByEmail` return whole user record, where `getUserByAccount` returns only selected fields

vilius commented on packages/adapter-drizzle-all/src/mysql/index.ts:
```diff
@@ -0,0 +1,293 @@
+        .then(res => res[0])
+    },
+    getUserByAccount: async (account) => {
+      return client.select({
```
Inconsistency: `getUser` and `getUserByEmail` returns whole user model where `getUserByAccount` returns only partial.

```ts
const userResult =
  client
    .select({
      user: usersTable,
    })
    .from(usersTable)
    .innerJoin(
      accountsTable,
      and(
        eq(accountsTable.providerAccountId, providerAccountId),
        eq(accountsTable.provider, provider),
      ),
    )
    .get() ?? null;

return userResult.then((userResult) => userResult.user);
```

anthonyshew replied:
Good catch! Forgot that the query returns both tables. 😄 Thank you!

ThangHuuVu commented on packages/adapter-drizzle-all/src/mysql/index.ts:
````diff
@@ -0,0 +1,293 @@
+ * ```
+ *
+ **/
+export function MySqlAdapter(client: typeof db): Adapter {
````
I think we need to have the schema objects passed in the Adapter, like this:
```suggestion
export function MySqlAdapter(client: typeof db, { users, accounts, sessions, verificationTokens  }: Schema): Adapter {
```

Then, we can get rid of the `schema.ts` file and `client.ts` file in the adapter code. WDYT? Normally, the userland `users` table is more complex than the standard `next-auth`'s `users` table

anthonyshew replied:
See https://github.com/nextauthjs/next-auth/pull/7165#issuecomment-1537254219

This has been updated for the other three flavored packages!

ThangHuuVu commented on packages/adapter-drizzle-all/src/mysql/index.ts:
```diff
@@ -0,0 +1,293 @@
+    useVerificationToken: async (token) => {
+      try {
+        await client
+          .delete(verificationTokens)
```
Shouldn't we select the token for later return first before deleting it?

anthonyshew replied:
Oops, yes!

ThangHuuVu commented on packages/adapter-drizzle-all/src/index.ts:
```diff
@@ -0,0 +1,26 @@
+import { PgAdapter } from "./pg"
+import { MySqlAdapter } from "./mysql"
```
should we get the other DB adapters from their packages?
```suggestion
import { MySqlAdapter } from "@next-auth/drizzle-adapter-mysql"
```

ThangHuuVu commented on packages/adapter-drizzle-pg/src/index.ts:
```diff
@@ -0,0 +1,264 @@
+    createUser: async (data) => {
+      return client
+        .insert(users)
+        .values({ ...data, id: "123" })
```
We need to get rid of these mock IDs

ThangHuuVu commented on packages/adapter-drizzle-pg/src/index.ts:
```diff
@@ -0,0 +1,264 @@
+
+      const account: ReturnType<Adapter["linkAccount"]> = {
+        ...updatedAccount,
+        access_token: updatedAccount.access_token ?? undefined,
```
Could you give this block of code some comments, it feels unnecessary to me 🤔

anthonyshew replied:
Added comment!
      // Drizzle will return `null` for fields that are not defined.
      // However, the return type is expecting `undefined`.

ThangHuuVu commented on packages/adapter-drizzle-sqlite/src/index.ts:
```diff
@@ -0,0 +1,261 @@
+    createUser: (data) => {
+      return client
+        .insert(users)
+        .values({ ...data, id: "123" })
```
we need to get rid of these mock Ids

balazsorban44 committed:
fix(docs): remove extra heading

Fixes #7426

anthonyshew committed:
merge main

anthonyshew committed:
Realigning structure.

anthonyshew committed:
Realigning structure.

anthonyshew committed:
Some hygiene.

anthonyshew committed:
Some hygiene.

balazsorban44 committed:
chore: use `@ts-ignore`

balazsorban44 committed:
chore: support release any package as experimental

balazsorban44 committed:
chore: separate manual release job

balazsorban44 committed:
chore: skip test for manual release

balazsorban44 committed:
chore: tweak

balazsorban44 committed:
chore: tweaks

balazsorban44 committed:
chore: tweak manual release version

anthonyshew committed:
Removing connection details for now.

anthonyshew committed:
Removing connection details for now.

anthonyshew committed:
Removing mock ids.

anthonyshew committed:
Removing mock ids.

anthonyshew committed:
Generate uuids.

anthonyshew committed:
Generate uuids.

anthonyshew committed:
Give name to compound keys.

anthonyshew committed:
Give name to compound keys.

anthonyshew committed:
Return the right thing from getUserByAccount.

anthonyshew committed:
Return the right thing from getUserByAccount.

anthonyshew committed:
Stop automatically migrating when generating types.

anthonyshew committed:
Stop automatically migrating when generating types.

anthonyshew committed:
Small cleanup.

anthonyshew committed:
Small cleanup.

anthonyshew committed:
More little cleanups.

anthonyshew committed:
More little cleanups.

anthonyshew committed:
Fix: grab token before deleting.

anthonyshew committed:
Fix: grab token before deleting.

anthonyshew committed:
Only use type definitions for packaged code.

anthonyshew committed:
Only use type definitions for packaged code.

anthonyshew committed:
Remove all adapter.

anthonyshew committed:
Remove all adapter.

anthonyshew commented:
After speaking with the Drizzle team, it sounds like their monorepo support is currently slightly unstable. I've opted to remove the "all"-style adapter because of this at this time. I'm figuring we can revisit this in the future if it turns out there is a need for such an adapter!

anthonyshew committed:
Merge branch 'main' into anthonyshew/main

anthonyshew committed:
Fix build scripts.

anthonyshew committed:
Fix build scripts.

anthonyshew committed:
Fix up test scripts.

anthonyshew committed:
Fix up test scripts.

anthonyshew committed:
Trying to fix test runs...?

anthonyshew committed:
Trying to fix test runs...?

anthonyshew commented:
@ThangHuuVu Did a clean up of your and others' reviews!

I'm having trouble getting tests to run at all this time around. Would you be able to help me with that?

balazsorban44 committed:
chore: reduce breaking changes on Account mapping

Reverts some changes on #7369 so DB migration won't be needed

balazsorban44 committed:
chore: add build to manual publish

balazsorban44 committed:
chore: move build to root

remirobichet committed:
docs(example): update broken link (#7504)

Co-authored-by: Nico Domino <yo@ndo.dev>

balazsorban44 committed:
feat: add update session to core (#7505)

* feat: add update session to core

Integrates #7056 into `@auth/core`

* resolve default user after jwt callback

oscartbeaumont commented:
To get the mysql adapter in this PR working with [Planetscale](https://planetscale.com) I had to change a couple of things.

First I had to remove `.references` as Planetscale doesn't support foreign key constraints. This can be done in userspace but does cause a Typescript mismatch.

I was also using the `@planetscale/database` serverless driver and this has a Typescript mismatch with the first argument to `DrizzleAdapterMySQL` as the DB type within this adapter is derived from a client using the `mysql` package instead.

Planetscale also doesn't seem to like the usage of MySQL `text` so I had to add the following alias into my Drizzle schema. This can be done in userspace but obviously causes a Typescript type error when passed to `DrizzleAdapterMySQL`.

```
const text = (name: string) =>
  varchar(name, {
    length: 255,
  });
```

I also had an issue with it crashing and have left that as a comment [here](https://github.com/nextauthjs/next-auth/pull/7165/files#diff-adab6ee4e101026d4b0131c568ceb9a01749b928bde501acef5d8acdf28bcb30R227).

Just wanted to document this as I think it will be fairly common for the MySql adapter to be used with Planetscale and the Planetscale serverless driver.

balazsorban44 committed:
chore: revert `picture` to `image`

anthonyshew commented:
Thanks for those notes, @oscartbeaumont! I had been wondering about compatibility up the dependency tree.

I'll yield to @ThangHuuVu and the core team here. How do y'all prefer to handle compatibility situations like these for adapters?

ghoshnirmalya committed:
chore: Move next.config.js file into the correct directory (#7580)

fix: moves next config file into the correct directory

roy9495 committed:
docs: Cypress.Cookies.defaults removed (#7574)

balazsorban44 committed:
feat: allow empty `account` mapper

balazsorban44 committed:
Merge branch 'main' of github.com:nextauthjs/next-auth

dorsha committed:
chore: Add Descope as a 🥉 bronze financial sponsor (#7615)

Add Descope as a bronze sponsor

wobsoriano committed:
chore(playgrounds): Nuxt 3.5.1 (#7626)

* bump Nuxt to 3.5.1

* follow playground package names

* chore: update nuxt playground scripts

* fix: imports and types

* fix: more nuxt type imports

* fix: nuxt auth options types

* fix: nuxt client fetch types

khuezy commented on packages/adapter-drizzle-mysql/src/index.ts:
```diff
@@ -0,0 +1,293 @@
+    linkAccount: async (rawAccount) => {
+      await client
+        .insert(accounts)
+        .values(rawAccount)
```
In the `accounts` schema: `expires_at`. But the rawAccount provided by the Google provider has `expires_in`. This will cause the insert to fail.
You'll probably want to pick out `expires_in` and add it to `expires_at: (Date.now()/1000 + expires_in)>>0`

anthonyshew replied:
Interesting, I was under the impression that the Github provider would handle this at its layer. 🤔  I'll be sure to be cognizant of this while I go around and test!

anthonyshew replied:
Going through testing now and it does look like this is working! Let me know if it isn't.

anthonyshew committed:
merge

anthonyshew committed:
Move back into one adapter.

anthonyshew committed:
Move back into one adapter.

anthonyshew committed:
With ugly TS now. Interface is too stiff.

anthonyshew committed:
With ugly TS now. Interface is too stiff.

anthonyshew committed:
Add PlanetScale.

anthonyshew committed:
Add PlanetScale.

anthonyshew committed:
Drop split up adapters.

anthonyshew committed:
Drop split up adapters.

anthonyshew committed:
Some cleanups.

anthonyshew committed:
Some cleanups.

anthonyshew commented:
For visibility:

- The core team has instructed me to opt for one adapter that handles all the SQL flavors.
- You'll also find support specifically for PlanetScale, paving a path for any other database providers that have quirks.
- I believe I've corrected TypeScript (in a messy way) such that users can now extend their schemas while still enforcing the minimum requirements of `next-auth`.
- I hope to take the `dist` out for a spin myself later this week and appreciate anyone else who goes and tests out stability!
- At this point, I still expect issues. 😀
- I have not handled documentation today as I still want to make sure that design is handled correctly.

pyrexfm commented:
The varchar(255) will lead to problems in the id_token field (happens with Google provider). Looking into [Prisma adapter for AuthJS](https://authjs.dev/reference/adapter/prisma), it seems that access_token, refresh_token and id_token should all be Text, so I changed and it works now.
However, I can't get it to work with Github provider.

`https://next-auth.js.org/errors#oauth_callback_handler_error Cannot read properties of undefined (reading 'name') TypeError: Cannot read properties of undefined (reading 'name')
    at eval (webpack-internal:///(sc_server)/./node_modules/drizzle-orm/session-6deff015.mjs:610:124)
    at Array.map (<anonymous>)
    at MySqlDialect.buildInsertQuery (webpack-internal:///(sc_server)/./node_modules/drizzle-orm/session-6deff015.mjs:610:40)
    at QueryPromise.getSQL (webpack-internal:///(sc_server)/./node_modules/drizzle-orm/session-6deff015.mjs:285:29)
    at QueryPromise.prepare (webpack-internal:///(sc_server)/./node_modules/drizzle-orm/session-6deff015.mjs:292:71)
    at MySqlInsert.execute (webpack-internal:///(sc_server)/./node_modules/drizzle-orm/session-6deff015.mjs:264:25)
    at QueryPromise.then (webpack-internal:///(sc_server)/./node_modules/drizzle-orm/relations-7adf6be2.mjs:1091:21)
    at linkAccount (webpack-internal:///(sc_server)/./src/lib/auth/drizzle-adapter.ts:58:108)
    at _callee2$ (webpack-internal:///(sc_server)/./node_modules/next-auth/core/errors.js:301:43)
    at tryCatch (webpack-internal:///(sc_server)/./node_modules/@babel/runtime/helpers/regeneratorRuntime.js:36:25)
    at Generator.eval (webpack-internal:///(sc_server)/./node_modules/@babel/runtime/helpers/regeneratorRuntime.js:120:30)
    at Generator.eval [as next] (webpack-internal:///(sc_server)/./node_modules/@babel/runtime/helpers/regeneratorRuntime.js:64:29)
    at asyncGeneratorStep (webpack-internal:///(sc_server)/./node_modules/@babel/runtime/helpers/asyncToGenerator.js:4:28)
    at _next (webpack-internal:///(sc_server)/./node_modules/@babel/runtime/helpers/asyncToGenerator.js:22:17)
    at eval (webpack-internal:///(sc_server)/./node_modules/@babel/runtime/helpers/asyncToGenerator.js:27:13)
    at new Promise (<anonymous>)
    at eval (webpack-internal:///(sc_server)/./node_modules/@babel/runtime/helpers/asyncToGenerator.js:19:16)
    at callbackHandler (webpack-internal:///(sc_server)/./node_modules/next-auth/core/lib/callback-handler.js:146:19)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async Object.callback (webpack-internal:///(sc_server)/./node_modules/next-auth/core/routes/callback.js:71:57)
    at async AuthHandler (webpack-internal:///(sc_server)/./node_modules/next-auth/core/index.js:202:38)
    at async NextAuthRouteHandler (webpack-internal:///(sc_server)/./node_modules/next-auth/next/index.js:49:30)
    at async NextAuth._args$ (webpack-internal:///(sc_server)/./node_modules/next-auth/next/index.js:83:24)
    at async eval (webpack-internal:///(sc_server)/./node_modules/next/dist/server/future/route-modules/app-route/module.js:242:37) {
  name: 'LinkAccountError',
  code: undefined`

howarddo2208 commented:
might be a dumb question, but how can I test out this adapter now before it's even merged? Thank u.

pyrexfm commented:
> might be a dumb question, but how can I test out this adapter now before it's even merged? Thank u.

Go to the src folder you need (pg, planetscale, mysql) and copy the schema and the index file into your project and create an adapter out of the index file. This allows you to experiment more easily.

anthonyshew committed:
Bug fixes.

anthonyshew committed:
Bug fixes.

anthonyshew committed:
Formatter pass.

anthonyshew committed:
Formatter pass.

khuezy commented on packages/adapter-drizzle/src/planetscale/index.ts:
```diff
@@ -0,0 +1,178 @@
+      }
+    },
+    deleteUser: async (id) => {
+      console.log({ id })
```
please remove this console.log

anthonyshew replied:
Oops, thank you! 😄

anthonyshew committed:
Remove console.log.

anthonyshew committed:
Remove console.log.

c-wide commented on packages/adapter-drizzle/src/mysql/index.ts:
```diff
@@ -0,0 +1,174 @@
+        .values(rawAccount)
+        .then((res) => res[0])
+    },
+    getUserByAccount: async (account) => {
```
Hello! Thank you for your hard work on this. 

I've been testing this out in a project I'm using next auth and drizzle for. I could be wrong here but I think this method will return the wrong user sometimes.

It appears to be selecting every user, performing a join, then returning the first user which will always be the first user registered in the database.

For example, I register an email account under 'test@example.com', log out and then register through an oauth provider that has a different email as the primary email (so a separate account is made), then log out and log back in through that same oauth provider you should be logged in as the first user in the database, not your oauth user. 

It's a bit of a convoluted example but it's the only one I've tested lol

Again thank you for your hard work so far!

khuezy replied:
Good point. This should:
1) query `accounts` table w/ `provider` and `providerAccountId`
2) If exists, query `users` by `account.id` to get user

anthonyshew replied:
Good catch! Thank you!

anthonyshew replied:
Yes, updated again!

anthonyshew committed:
Documentation!

anthonyshew committed:
Documentation!

anthonyshew committed:
PlanetScale needs varchar but types are unhappy...

anthonyshew committed:
PlanetScale needs varchar but types are unhappy...

mcnaveen commented:
Wow, Thanks guys for working on it. 🎉  Looking forward to it!

anthonyshew committed:
Correct some types.

anthonyshew committed:
Correct some types.

anthonyshew committed:
Fix logical bug for returning wrong account.

anthonyshew committed:
Fix logical bug for returning wrong account.

anthonyshew committed:
Trying to get tests to run...

anthonyshew committed:
Trying to get tests to run...

anthonyshew commented:
Another update!:
- Manually tested the Postgres and Planetscale adapters. They appear to be working correctly for me. They still worked as expected when I added additional fields to the schema.
- Updated the API interface so that you don't have to provide a magic string for your SQL flavor.
- Wrote documentation.

Outstanding issues (that I'd probably need help resolving):
- Still can't get the tests to run! Drizzle resolves fine everywhere but in the Jest tests. 🤔 I've scaffolded out the tests for Postgres if anyone wants to try their hand finding the fix.
- The way I have TypeScript set up, it doesn't enforce the minimum schema requirements. This does leave some room for error for users - but I'm not certain about how to describe the types more succinctly. I'm not sure about the strictness of the other adapters so I'm not sure if this is to be considered blocking or not.

Additionally, if anyone wants to keep trying it out, feel free to report bugs! 😄

We're close!

anthonyshew committed:
Update package imports.

anthonyshew committed:
Update package imports.

anthonyshew committed:
Make one query in MySQL.

anthonyshew committed:
Make one query in MySQL.

balazsorban44 committed:
revert unrelated changes

balazsorban44 committed:
update lock file

balazsorban44 committed:
restructure, ESM-only

balazsorban44 committed:
fix imports

balazsorban44 committed:
recognize mjs and cjs in jest

balazsorban44 commented:
Fixed merge conflicts and the tests. `drizzle-orm` only publishes `.mjs` and `.cjs` files and our jest config ignored those.

balazsorban44 committed:
revert package.json

balazsorban44 committed:
remove uuid, format

balazsorban44 committed:
revert some changes

anthonyshew committed:
Merge branch 'main' of https://github.com/anthonyshew/next-auth into anthonyshew/main

Lermatroid commented:
Hi, I'm attempting to use this at the moment but I seem to be running into a type error when setting the adapter.

Here is my next auth config:

```ts
export const authOptions: NextAuthOptions = {
  callbacks: {
    session: ({ session, user }) => ({
      ...session,
      user: {
        ...session.user,
        id: user.id,
      },
    }),
  },
  adapter:  DrizzleAdapter(db, { users, accounts, sessions, verificationTokens }),
  providers: [
    DiscordProvider({
      clientId: env.DISCORD_CLIENT_ID,
      clientSecret: env.DISCORD_CLIENT_SECRET,
    }),
    /**
     * ...add more providers here.
     *
     * Most other providers require a bit more work than the Discord provider. For example, the
     * GitHub provider requires you to add the `refresh_token_expires_in` field to the Account
     * model. Refer to the NextAuth.js docs for the provider you want to use. Example:
     *
     * @see https://next-auth.js.org/providers/github
     */
  ],
};
```

The error is as follows:

```
Type 'PgTableWithColumns<{ name: "verificationToken"; schema: undefined; columns: { identifier: PgText<{ tableName: "verificationToken"; data: string; enumValues: [string, ...string[]]; name: "identifier"; driverParam: string; hasDefault: false; notNull: true; }>; token: PgText<...>; expires: PgInteger<...>; }; }>' is not assignable to type 'PgTableWithColumns<{ name: "verificationToken"; schema: undefined; columns: { identifier: PgText<{ tableName: "verificationToken"; data: string; enumValues: [string, ...string[]]; name: "identifier"; driverParam: string; hasDefault: false; notNull: true; }>; token: PgText<...>; expires: PgTimestamp<...>; }; }>'.
  Type 'PgTableWithColumns<{ name: "verificationToken"; schema: undefined; columns: { identifier: PgText<{ tableName: "verificationToken"; data: string; enumValues: [string, ...string[]]; name: "identifier"; driverParam: string; hasDefault: false; notNull: true; }>; token: PgText<...>; expires: PgInteger<...>; }; }>' is not assignable to type 'PgTable<{ name: "verificationToken"; schema: undefined; columns: { identifier: PgText<{ tableName: "verificationToken"; data: string; enumValues: [string, ...string[]]; name: "identifier"; driverParam: string; hasDefault: false; notNull: true; }>; token: PgText<...>; expires: PgTimestamp<...>; }; }>'.
    The types of '_.config.columns.expires' are incompatible between these types.
      Type 'PgInteger<{ tableName: "verificationToken"; data: number; name: "expires"; driverParam: string | number; hasDefault: false; notNull: true; }>' is missing the following properties from type 'PgTimestamp<{ tableName: "verificationToken"; data: Date; name: "expires"; driverParam: string; hasDefault: false; notNull: true; }>': withTimezone, precisionts(2322)
schema.ts(73, 3): The expected type comes from property 'verificationTokens' which is declared here on type 'Schema & Record<string, AnyPgTable>'
```

Any idea what could be causing this?

anthonyshew committed:
Checking in, sleepy time.

mcnaveen commented on packages/adapter-drizzle/src/mysql/index.ts:
```diff
@@ -0,0 +1,252 @@
+  process.env.AUTH_DRIZZLE_MYSQL_URL as string,
+) : createPool({
+  host: "localhost",
+  user: "root",
```
Is this supposed to be coming from the env right?

XMadFoX commented:
Hi @anthonyshew, thank you for working on this adapter.

I've been testing it and faced a problem with implementing [credentials provider](https://next-auth.js.org/providers/credentials) (username/password example is simple reproduction), (email provider works fine).
As I understand, after calling `authorize` and returning user object from there, a session should be created in database.

But debugging network & adapter shows:
After authorize called, 302 /credentials returns next-auth.session-token (jwt)
Then comes request to /session (200) which deletes token from cookie (returns value `""`)
as `getSessionAndUser(token)` returns null, as it cannot find session (it's not being created as mentioned above)

Database is sqlite/turso, commenting out adapter in authOptions => credentials provider works as expected

Could you please look into that?

anthonyshew committed:
Able to run tests again! Bumped drizzle back to 0.24.x

matheins commented on packages/adapter-drizzle/src/planetscale/index.ts:
```diff
@@ -0,0 +1,248 @@
+          .leftJoin(users, eq(accounts.userId, users.id))
+          .then(res => res[0])
+
+      return dbAccount.users
```
Returning the query result is causing an Callback error, in one edge case:
- If user already exists but has no linked oauth account and tries to sign in

This check before returning solves it: 
`
if (!dbAccount) return null
`

matheins commented on packages/adapter-drizzle/src/planetscale/index.ts:
```diff
@@ -0,0 +1,248 @@
+    token_type: varchar("token_type", { length: 255 }),
+    scope: varchar("scope", { length: 255 }),
+    id_token: varchar("id_token", { length: 255 }),
+    session_state: varchar("session_state", { length: 255 }),
```
`linkAccount` function fails due to missing prop in the schema, when trying to register via github:
`refresh_token_expires_in: int("refresh_token_expires_in"),`

PatrickJoannisse replied:
I got an issue related to linkAccount as well but the issue was that id_token's length was not long enough for some providers. Notably Google and Twitch. I switched it to a Text type which is probably overkill but it works.

nolanholden commented:
@anthonyshew -- Do you forsee instability using this branch with Planetscale as-is? Considering using it as a holdover until your merge goes through.

anthonyshew commented:
@nolanholden Yes, I was working on this without being able to run tests for awhile so I've got a lot to circle back on fixing. 😄

andresgutgon commented on packages/adapter-drizzle/src/pg/index.ts:
```diff
@@ -0,0 +1,238 @@
+        .delete(users)
+        .where(eq(users.id, id))
+        .returning()
+        .then((res) => res[0])
```
I see this when I copy this file in my code
<img width="1212" alt="image" src="https://github.com/nextauthjs/next-auth/assets/49499/effcd41f-90d6-4e82-a7b2-17f2c6ec8234">


Shouldn't be `res.rows[0]`?

andresgutgon commented on packages/adapter-drizzle/src/pg/index.ts:
```diff
@@ -0,0 +1,238 @@
+        .insert(users)
+        .values({ ...data, id: crypto.randomUUID() })
+        .returning()
+        .then((res) => res[0])
```
This is giving me a typescript error when trying to pass the adapter to `nextauth` as Adapter

```
Type 'Adapter' is not assignable to type 'Adapter | undefined'.
  Type 'Adapter' is not assignable to type 'DefaultAdapter & { createVerificationToken: (verificationToken: VerificationToken) => Awaitable<VerificationToken | null | undefined>; useVerificationToken: (params: { ...; }) => Awaitable<...>; }'.
    Type 'Adapter' is not assignable to type 'DefaultAdapter'.
      Types of property 'createUser' are incompatible.
        Type '((user: Omit<AdapterUser, "id">) => Awaitable<AdapterUser>) | undefined' is not assignable to type '(user: Omit<AdapterUser, "id">) => Awaitable<AdapterUser>'.
          Type 'undefined' is not assignable to type '(user: Omit<AdapterUser, "id">) => Awaitable<AdapterUser>'.

```

andresgutgon replied:
This is giving me a typescript error when trying to pass the adapter to `nextauth` as Adapter

```
Type 'Adapter' is not assignable to type 'Adapter | undefined'.
  Type 'Adapter' is not assignable to type 'DefaultAdapter & { createVerificationToken: (verificationToken: VerificationToken) => Awaitable<VerificationToken | null | undefined>; useVerificationToken: (params: { ...; }) => Awaitable<...>; }'.
    Type 'Adapter' is not assignable to type 'DefaultAdapter'.
      Types of property 'createUser' are incompatible.
        Type '((user: Omit<AdapterUser, "id">) => Awaitable<AdapterUser>) | undefined' is not assignable to type '(user: Omit<AdapterUser, "id">) => Awaitable<AdapterUser>'.
          Type 'undefined' is not assignable to type '(user: Omit<AdapterUser, "id">) => Awaitable<AdapterUser>'.

```

andresgutgon commented on packages/adapter-drizzle/src/mysql/index.ts:
```diff
@@ -0,0 +1,254 @@
+        .select()
+        .from(users)
+        .where(eq(users.id, id))
+        .then((res) => res[0])
```
This is giving me a typescript error when trying to pass the adapter to `nextauth` as Adapter

```
Type 'Adapter' is not assignable to type 'Adapter | undefined'.
  Type 'Adapter' is not assignable to type 'DefaultAdapter & { createVerificationToken: (verificationToken: VerificationToken) => Awaitable<VerificationToken | null | undefined>; useVerificationToken: (params: { ...; }) => Awaitable<...>; }'.
    Type 'Adapter' is not assignable to type 'DefaultAdapter'.
      Types of property 'createUser' are incompatible.
        Type '((user: Omit<AdapterUser, "id">) => Awaitable<AdapterUser>) | undefined' is not assignable to type '(user: Omit<AdapterUser, "id">) => Awaitable<AdapterUser>'.
          Type 'undefined' is not assignable to type '(user: Omit<AdapterUser, "id">) => Awaitable<AdapterUser>'.

```

andresgutgon replied:
I fixed the error by lying Typescript, not even tried yet. But I would like to understand why `.where` is giving `undefined` in Drizzel. Please comment if you know why : )

<img width="540" alt="image" src="https://github.com/nextauthjs/next-auth/assets/49499/1b9e6d79-2cea-43ad-a2b8-2452cef9baa8">

andresgutgon commented on packages/adapter-drizzle/src/mysql/index.ts:
```diff
@@ -0,0 +1,254 @@
+  {
+    userId: varchar("userId", { length: 255 })
+      .notNull()
+      .references(() => users.id, { onDelete: "cascade" }),
```
Planet scale does not support foreign keys. So I had to remove `.references` to be able to push the migrations to their servers.

<img width="1290" alt="image" src="https://github.com/nextauthjs/next-auth/assets/49499/0d4da236-b2a8-4b96-9ad3-5edbb5e37847">

PatrickJoannisse replied:
I'm also using Planetscale but have you seen this `packages/adapter-drizzle/src/planetscale/index.ts`?

andresgutgon replied:
It's what I'm using. The problem is I think that Planet Scale use a flavor of MySQL called Vitess which is mean to be scalable and they don't allow foreign keys. That's my theory. 

Any way this issue goes away not using `.references`. You don't have foreign keys but that's the price for using something at planet scale (pum intended)

VaniaPopovic replied:
There are two seperate adapters, one for mysql and one for planetscale in the PR.

andresgutgon replied:
Oh you're right my bad. Ignore my comment here

khuezy commented on packages/adapter-drizzle/src/sqlite/index.ts:
```diff
@@ -0,0 +1,223 @@
+      return account
+    },
+    getUserByAccount: (account) => {
+      const dbAccount =
```
```suggestion
      const results = await client
       .select().from(accounts)
       .leftJoin(users, eq(users.id, accounts.id))
       .where(
         and(eq(accounts.provider, account.provider), eq(accounts.providerAccountId, account.providerAccountId))
       ).get()
       
       return results.users ?? null;
      
```

michaelangrivera replied:
This is likely more performant. I don't see any indexes made in this PR. Is that something that should be done as well?

michaelangrivera replied:
https://github.com/nextauthjs/next-auth/pull/7165/files#r1253130773 index suggestions here. There should be indexes for each type (pg, MySQL, sqlite) IMO

anthonyshew replied:
Thanks!

anthonyshew committed:
MySQL tests are passing...

anthonyshew committed:
Both MySQL tests work.

nolanholden commented:
Excited for this one! @anthonyshew

anthonyshew committed:
Passes for MySQL and setting up other tests.

andresgutgon commented on packages/adapter-drizzle/src/planetscale/index.ts:
```diff
@@ -0,0 +1,248 @@
+      await client
+        .insert(accounts)
+        .values(rawAccount)
+        .then((res) => res.rows[0])
```
This is failing for me trying Google Auth. The `rawAccount.id_token` is bigger than 255 chars. Is 1185 chars. So maybe this column `id_token` should be a text column.

anthonyshew replied:
Will be dropping PlanetScale to cut scope so we can ship. We should be able to add PlanetScale back in the future!

anthonyshew committed:
SQLite tests pass!

anthonyshew committed:
Adjust Postgres tests for correctness.

anthonyshew committed:
All tests are passing.

michaelangrivera commented on packages/adapter-drizzle/src/sqlite/index.ts:
```diff
@@ -0,0 +1,223 @@
+  },
+  (account) => ({
+    compoundKey: primaryKey(account.provider, account.providerAccountId),
+  })
```
Indexes should be made for this table, likely on the following columns:
- `userId` -> uniqueIndex
- `type` -> index
- `provider` -> index
- `expires_at` -> index (descending?) 
- `session_state` -> index

michaelangrivera replied:
https://orm.drizzle.team/docs/indexes-constraints#indexes

khuezy replied:
The adapter's access pattern is only user.id and user.email. For `accounts` it's only by `provider and providerAccountId`, so we don't need indexes for those fields unless you're doing something specific out of scope of the Adapter.

michaelangrivera replied:
Perhaps it would be nice to give the option of accessing these?

anthonyshew replied:
Users will be able to provide custom tables; you'll be able to add your own custom indices.

Do we think any of these suggestions should be there by default? Going to mark as resolved because it doesn't look like like there should be at a glance?

arivera-xealth replied:
@anthonyshew I think at the very least the indexes should be mentioned in the doc. When you scale to millions of users, this is a big deal.

anthonyshew replied:
@arivera-xealth We need to ensure we're careful about scope here. The docs need to describe what makes sense for the default configuration. If users are creating extra access patterns such that indices make sense for their application, that doesn't seem like something we need to document in next-auth.

For the default usage of next-auth, do we need to add any extra indices? From what I can tell looking at prior art (namely, the Prisma adapter), it doesn't look like we need to.

andresgutgon replied:
My 2 cents. I would follow previous implementations on other adapters. Add just the indices those adapters added

michaelangrivera commented on packages/adapter-drizzle/src/pg/index.ts:
```diff
@@ -0,0 +1,244 @@
+    providerAccountId: text("providerAccountId").notNull(),
+    refresh_token: text("refresh_token"),
+    access_token: text("access_token"),
+    expires_at: integer("expires_at"),
```
why not use the native PG date column? Let the database do the work, not the consumer. 

https://orm.drizzle.team/docs/column-types/pg#timestamp

anthonyshew replied:
Need to follow the spec: https://authjs.dev/reference/adapters#models

michaelangrivera commented on packages/adapter-drizzle/src/index.ts:
````diff
@@ -0,0 +1,355 @@
+ * ```
+ *
+ **/
+export function DrizzleAdapter<SqlFlavor extends SqlFlavorOptions>(
````
What is the plan for migrations?

anthonyshew committed:
Fixing up TS.

anthonyshew committed:
Merge branch 'main' of https://github.com/nextauthjs/next-auth

anthonyshew committed:
update lock

anthonyshew committed:
Remove connection info from source.

anthonyshew committed:
Need container to cooperate...

anthonyshew committed:
All tests passing again!

anthonyshew committed:
Merge pull request #2 from anthonyshew/what

What

anthonyshew committed:
Passing all zero-config.

anthonyshew committed:
Setting up but not executing custom tests.

anthonyshew commented on packages/adapter-drizzle/src/sqlite/index.ts:
```diff
@@ -0,0 +1,223 @@
+    token_type: text("token_type"),
+    scope: text("scope"),
+    id_token: text("id_token"),
+    session_state: text("session_state"),
```
We need to follow the spec: https://authjs.dev/reference/adapters#models

Users will be able to customize the tables if they want.

anthonyshew committed:
Dropping PlanetScale to cut scope.

anthonyshew committed:
Merge pull request #3 from anthonyshew/what

What

anthonyshew committed:
Perf tweak.

anthonyshew committed:
Merge pull request #4 from anthonyshew/what

Perf tweak.

nolanholden commented:
hey @anthonyshew, not attacking the PR here, but I am curious. Why are there 189 commits? Is this because you are merging main back into this branch without ever rebasing? Could some commits be squashed? I know if I was a maintainer I would be concerned about this.

But there may be something I don't know! Interested about your process.

anthonyshew commented:
@nolanholden I'm going to do some rebasing and squashing once I'm done; just haven't cared to do it yet. 😁 We've changed the API we were going for like four times and keeping the prior art around has helped a few times.

anthonyshew committed:
Docced and everything!

anthonyshew committed:
Remove unused exports.

anthonyshew committed:
Merge pull request #5 from anthonyshew/what

What

anthonyshew committed:
Add fixtures to workaround millisecond precision.

anthonyshew committed:
Merge pull request #6 from anthonyshew/what

Add fixtures to workaround millisecond precision.

anthonyshew committed:
Fix lints.

anthonyshew committed:
Run prettier.

anthonyshew committed:
Merge pull request #7 from anthonyshew/what

Lint and format.

anthonyshew committed:
Merge branch 'main' into main

anthonyshew commented:
Okay, folks, I think we've got it! I've written up tests and manually tested all three SQL flavors with Docker.

One thing of note: I would recommend marking this adapter as experimental. 
- Through the course of working on this, Drizzle shipped several breaking changes that affected both the correctness and development velocity of this PR. Don't read this as a critique; rather, this would be an acknowledgement that library is still quite young (despite its popularity). The Drizzle core team is *excellent* and their ORM will be stable and awesome in the future - but it's important to note the downstream dependency, in my opinion.

andresgutgon commented on packages/adapter-drizzle/src/mysql/index.ts:
```diff
@@ -0,0 +1,261 @@
+    token_type: varchar("token_type", { length: 255 }),
+    scope: varchar("scope", { length: 255 }),
+    id_token: varchar("id_token", { length: 255 }),
+    session_state: varchar("session_state", { length: 255 }),
```
I copied this code in my project because I wanted to start using Planetscale + Next-Auth + Drizzel. So thank you!. 

Is this mysql adapter the one we should use if we want to use Planetscale? If yes, in my case I had to use for `id_token` and `session_state` a `text` field not a `varchar` because I got an error when using Google Auth.

andresgutgon replied:
Sorry, I didn't see your previous comment here https://github.com/nextauthjs/next-auth/pull/7165#discussion_r1263313311

Ignore this. I guess I'll use my custom adapter for planet scale for now. Thanks for the work!

marvkr replied:
Would you mind sharing your adapter? I'm also using Planetscale so this bugs me as well ... 😕

demarchenac replied:
Hi @andresgutgon! It would be nice, maybe you have a public repo in which we can read your twist of the adapter?

iukea1 replied:
> I copied this code in my project because I wanted to start using Planetscale + Next-Auth + Drizzel. So thank you!. 
> 
> Is this mysql adapter the one we should use if we want to use Planetscale? If yes, in my case I had to use for `id_token` and `session_state` a `text` field not a `varchar` because I got an error when using Google Auth.
> 
> 
Sharing is caring

andresgutgon replied:
Not much to it. I just copy & paste from this PR when there was the planetscale adapter and fixed the errors for Google Auth provider.
This is working for me: Google Auth + Planetscale
https://gist.github.com/andresgutgon/dcbf135f890450f0a03eeb9b7891b268

demarchenac replied:
I was just wondering since I was looking at the schema generated for Primsa, but shouldn't the `refresh_token` & `access_token` columns have a `text` type as well?

andresgutgon replied:
In my case with Google Auth was not necessary. The problem with using text instead of varchar is that it takes more space in the db to store the column afaik. So i use text only when strictly necessary

iukea1 replied:
@andresgutgon  did you get prepared queries working?

andresgutgon replied:
I don't even know what's that lol. So I guess no xD. What you see in the gist is what I did. Nothing else related with Drizzle + NextAuth + Planetscale

iukea1 replied:
@andresgutgon  prepared queries with your build are not erroring out locally. Pushing to Versel now

andresgutgon replied:
Can you explain here quickly what's a prepared query and what's the benefit?

mschieller replied:
A prepared statement is kinda what the name implies. Instead of generate each sql query on request which is the normal procedure usually when using ORMs, Drizzle can make prepared statements, where the SQL is made on build, and when requested by code it it can use it directly which improves performance a lot because it can reuse the binary compilation. It can also take in params if doing joins and what not, which still is much faster. See here for more https://orm.drizzle.team/docs/performance#

iukea1 replied:
so Prepared statement

> When you run query to the database there're several things that happens:
> 
> - all the configurations of the query builder got concatenated to the SQL string
>-  that string and params are sent to the database driver
>-  driver compiles SQL query to the binary SQL executable format and sends it to the database
>-  With prepared statements you do SQL concatenation once on the Drizzle ORM side and then database driver is able to reuse precompiled binary SQL instead of parsing query all the time. It has extreme performance benefits on large SQL queries.

https://orm.drizzle.team/docs/performance

can also watch the first half of this video
- [](https://www.youtube.com/watch?v=O9zxsSl9zgE&t=430s)

michaelangelo-io commented on packages/adapter-drizzle/src/pg/index.ts:
```diff
@@ -0,0 +1,231 @@
+        .returning()
+        .then((res) => res[0])
+    },
+    linkAccount: async (rawAccount) => {
```
If you get the following error with Github provider:
```
      "message": "Cannot read properties of undefined (reading 'name')",
      "stack": "TypeError: Cannot read properties of undefined (reading 'name')\n
```
It's because GitHub adds an extra property `refresh_token_expires_in`: 
```typescript
{
  provider: 'github',
  type: 'oauth',
  providerAccountId: '',
  access_token: '',
  expires_at: 1690073663,
  refresh_token: '',
  refresh_token_expires_in: 15897600, // <----- 
  token_type: 'bearer',
  scope: '',
  userId: '53c777ed-353a-438b-a57e-wqweqwe'
}
```
I think maybe it would be best (for compatibility) to add this column to the provider? Or I guess maybe point that out in the docs so people can add that column to the schema? 

Also, it may be better to use bigInt (with mode `number`) instead of integer for the expires column(s) for flexibility. https://orm.drizzle.team/docs/column-types/pg#bigint

mschieller commented:
Is there a reason to why this adapter is not using prepared statements? Hence prepared statements has a much quicker build time that standard typings, and that these auth request would be called millions of millions of time, prepared statements should be really beneficial?
And if decided not to use prepared as default, perhaps make it optional in the adapter, or would this change the decires outcome of the generality of all the adapters?

andresgutgon commented:
> I created a fully functional NextAuth adapter for Drizzle+PlanetScale, which utilises the new query methodologies, optional prepared statements to be used in adapter. If any would want it, I can share my findings and code :-)

Yes please, I [shared mine here](https://github.com/nextauthjs/next-auth/pull/7165#discussion_r1278520458) but not with prepare statements. I'm curious

mschieller commented:
> > I created a fully functional NextAuth adapter for Drizzle+PlanetScale, which utilises the new query methodologies, optional prepared statements to be used in adapter. If any would want it, I can share my findings and code :-)
> 
> 
> 
> Yes please, I [shared mine here](https://github.com/nextauthjs/next-auth/pull/7165#discussion_r1278520458) but not with prepare statements. I'm curious

Currently flying home from vacation, but I'll make a test-next repo tonight and post it here, so all of it can be seen 🫡

mschieller commented:
> > > I created a fully functional NextAuth adapter for Drizzle+PlanetScale, which utilises the new query methodologies, optional prepared statements to be used in adapter. If any would want it, I can share my findings and code :-)
> > 
> > 
> > Yes please, I [shared mine here](https://github.com/nextauthjs/next-auth/pull/7165#discussion_r1278520458) but not with prepare statements. I'm curious
> 
> Currently flying home from vacation, but I'll make a test-next repo tonight and post it here, so all of it can be seen 🫡

**As promised** https://github.com/mschieller/drizzle-nextauth-planetscale
Feel free to ask why, and if you see something wrong, please let me know.

be aware that typing was a bit tricky so the adapter is not typesafe yet

zivtamary commented:
> **As promised** https://github.com/mschieller/drizzle-nextauth-planetscale Feel free to ask why, and if you see something wrong, please let me know.
> 
> be aware that typing was a bit tricky so the adapter is not typesafe yet

Hey, thanks for that :) would it also work for database session strategy?

mschieller commented:
> > **As promised** https://github.com/mschieller/drizzle-nextauth-planetscale Feel free to ask why, and if you see something wrong, please let me know.
> 
> > 
> 
> > be aware that typing was a bit tricky so the adapter is not typesafe yet
> 
> 
> 
> Hey, thanks for that :) would it also work for database session strategy?

Pretty confident it does yes, haven't tried though. Please give feedback if/when you have tried 😀

zivtamary commented:
> > > **As promised** https://github.com/mschieller/drizzle-nextauth-planetscale Feel free to ask why, and if you see something wrong, please let me know.
> > 
> > 
> > > 
> > 
> > 
> > > be aware that typing was a bit tricky so the adapter is not typesafe yet
> > 
> > 
> > Hey, thanks for that :) would it also work for database session strategy?
> 
> Pretty confident it does yes, haven't tried though. Please give feedback if/when you have tried 😀
Unfortunately getting some errors
`[next-auth][warn][EXPERIMENTAL_API] 
`getServerSession` is used in a React Server Component. 
https://next-auth.js.org/configuration/nextjs#getServerSession}
https://next-auth.js.org/warnings#EXPERIMENTAL_API
[next-auth][error][adapter_error_linkAccount] 
https://next-auth.js.org/errors#adapter_error_linkaccount Cannot read properties of undefined (reading 'name') {
  message: "Cannot read properties of undefined (reading 'name')",
  stack: "TypeError: Cannot read properties of undefined (reading 'name')\n" +
    '    at eval (webpack-internal:///(rsc)/./node_modules/drizzle-orm/session-9628aea0.mjs:727:127)\n' +
    '    at Array.map (<anonymous>)\n' +
    '    at MySqlDialect.buildInsertQuery (webpack-internal:///(rsc)/./node_modules/drizzle-orm/session-9628aea0.mjs:727:40)\n' +
    '    at QueryPromise.getSQL (webpack-internal:///(rsc)/./node_modules/drizzle-orm/session-9628aea0.mjs:383:29)\n' +
    '    at QueryPromise.prepare (webpack-internal:///(rsc)/./node_modules/drizzle-orm/session-9628aea0.mjs:390:71)\n' +
    '    at MySqlInsert.execute (webpack-internal:///(rsc)/./node_modules/drizzle-orm/session-9628aea0.mjs:360:25)\n' +
    '    at QueryPromise.then (webpack-internal:///(rsc)/./node_modules/drizzle-orm/alias-3e926a50.mjs:766:21)\n' +
    '    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)',
  name: 'TypeError'
}
[next-auth][error][OAUTH_CALLBACK_HANDLER_ERROR] 
https://next-auth.js.org/errors#oauth_callback_handler_error Cannot read properties of undefined (reading 'name') TypeError: Cannot read properties of undefined (reading 'name')
    at eval (webpack-internal:///(rsc)/./node_modules/drizzle-orm/session-9628aea0.mjs:727:127)
    at Array.map (<anonymous>)
    at MySqlDialect.buildInsertQuery (webpack-internal:///(rsc)/./node_modules/drizzle-orm/session-9628aea0.mjs:727:40)
    at QueryPromise.getSQL (webpack-internal:///(rsc)/./node_modules/drizzle-orm/session-9628aea0.mjs:383:29)
    at QueryPromise.prepare (webpack-internal:///(rsc)/./node_modules/drizzle-orm/session-9628aea0.mjs:390:71)
    at MySqlInsert.execute (webpack-internal:///(rsc)/./node_modules/drizzle-orm/session-9628aea0.mjs:360:25)
    at QueryPromise.then (webpack-internal:///(rsc)/./node_modules/drizzle-orm/alias-3e926a50.mjs:766:21)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5) {
  name: 'LinkAccountError',
  code: undefined
}`

zivtamary commented:
> > > **As promised** https://github.com/mschieller/drizzle-nextauth-planetscale Feel free to ask why, and if you see something wrong, please let me know.
> > 
> > 
> > > 
> > 
> > 
> > > be aware that typing was a bit tricky so the adapter is not typesafe yet
> > 
> > 
> > Hey, thanks for that :) would it also work for database session strategy?
> 
> Pretty confident it does yes, haven't tried though. Please give feedback if/when you have tried 😀

Unfortunately getting some errors... btw I'm using Google adapter
```
[next-auth][error][adapter_error_linkAccount] 
https://next-auth.js.org/errors#adapter_error_linkaccount Cannot read properties of undefined (reading 'name') {
  message: "Cannot read properties of undefined (reading 'name')",
  stack: "TypeError: Cannot read properties of undefined (reading 'name')\n" +
```

mschieller commented:
> > > > **As promised** https://github.com/mschieller/drizzle-nextauth-planetscale Feel free to ask why, and if you see something wrong, please let me know.
> 
> > > 
> 
> > > 
> 
> > > > 
> 
> > > 
> 
> > > 
> 
> > > > be aware that typing was a bit tricky so the adapter is not typesafe yet
> 
> > > 
> 
> > > 
> 
> > > Hey, thanks for that :) would it also work for database session strategy?
> 
> > 
> 
> > Pretty confident it does yes, haven't tried though. Please give feedback if/when you have tried 😀
> 
> 
> 
> Unfortunately getting some errors... btw I'm using Google adapter
> 
> ```
> 
> [next-auth][error][adapter_error_linkAccount] 
> 
> https://next-auth.js.org/errors#adapter_error_linkaccount Cannot read properties of undefined (reading 'name') {
> 
>   message: "Cannot read properties of undefined (reading 'name')",
> 
>   stack: "TypeError: Cannot read properties of undefined (reading 'name')\n" +
> 
> ```

I got the same errors my self, forgot to redo a change I had made as a test prior. Check the new update I have made to the auth schema :-)

zivtamary commented:
> > > > > **As promised** https://github.com/mschieller/drizzle-nextauth-planetscale Feel free to ask why, and if you see something wrong, please let me know.
> > 
> > 
> > > >
> > 
> > 
> > > >
> > 
> > 
> > > > >
> > 
> > 
> > > >
> > 
> > 
> > > >
> > 
> > 
> > > > > be aware that typing was a bit tricky so the adapter is not typesafe yet
> > 
> > 
> > > >
> > 
> > 
> > > >
> > 
> > 
> > > > Hey, thanks for that :) would it also work for database session strategy?
> > 
> > 
> > > 
> > 
> > 
> > > Pretty confident it does yes, haven't tried though. Please give feedback if/when you have tried 😀
> > 
> > 
> > Unfortunately getting some errors... btw I'm using Google adapter
> > ```
> > 
> > [next-auth][error][adapter_error_linkAccount] 
> > 
> > https://next-auth.js.org/errors#adapter_error_linkaccount Cannot read properties of undefined (reading 'name') {
> > 
> >   message: "Cannot read properties of undefined (reading 'name')",
> > 
> >   stack: "TypeError: Cannot read properties of undefined (reading 'name')\n" +
> > ```
> 
> I got the same errors my self, forgot to redo a change I had made as a test prior. Check the new update I have made to the auth schema :-)

It works like a charm! Thank you! 😀

balazsorban44 commented:
Hi folks, this is now released under `@auth/drizzle-adapter`! Please try it out. 🙏

Docs: https://authjs.dev/reference/adapter/drizzle

Note, it should work with both NextAuth.js and Auth.js libs.

Thanks for everyone chipping in and your patience! Let us know!
